FLT/AutomorphicForm/QuaternionAlgebra/FiniteDimensional.lean

```diff
@@ -1,31 +1,124 @@
+-- let F be a number field
+variable {F : Type*} [Field F] [NumberField F]
     -- and let D be a totally definite quaternion algebra over F
-    {D : Type*} [Ring D] [Algebra F D] [IsQuaternionAlgebra F D]
+    -- TODO the assumption that D is totally definite implies that
+    -- D is a division ring, so we should be able to remove the
+    -- [DivisionRing D] requirement.
+    {D : Type*} [DivisionRing D] [Algebra F D] [IsQuaternionAlgebra F D]
     (hD : IsTotallyDefinite F D)
-
 -- Let K be a coefficient field
 variable (K : Type*) [Field K]
     -- and let U, the level, be a (compact) open subgroup of `(D ⊗ 𝔸_F^∞)ˣ`
-    (U : Subgroup (D ⊗[F] (FiniteAdeleRing (𝓞 F) F))ˣ)
-    (hU : IsOpen (U : Set (D ⊗[F] (FiniteAdeleRing (𝓞 F) F))ˣ))
+    (U : Subgroup (Dfx (F:=F) (D:=D)))
+    (hU : IsOpen (U : Set (Dfx (F:=F) (D:=D))))
 
 open TotallyDefiniteQuaternionAlgebra
 
+
 -- Then the space of K-valued weight 2 level U modular forms for Dˣ is finite-dimensional over K.
 /-- The space of `K`-valued weight 2 level `U` quaternionic automorphic forms
 for `Dˣ` is finite-dimensional over `K`.
 -/
-theorem WeightTwoAutomorphicForm.finiteDimensional :
-    FiniteDimensional K (WeightTwoAutomorphicFormOfLevel U K) :=
-  sorry
+theorem WeightTwoAutomorphicForm.finiteDimensional
+    {hU : IsOpen (U : Set (Dfx (F:=F) (D:=D)))} :
+    FiniteDimensional K (WeightTwoAutomorphicFormOfLevel U K) := by
+    let U' : Subgroup (Dfx F D) := U
+    let H' : Subgroup (Dfx F D) := {
+        carrier := Set.range (incl₁ F D),
+        one_mem' := ⟨1, by simp⟩,
+        mul_mem' := by
+            rintro _ _ ⟨a, rfl⟩ ⟨b, rfl⟩
+            exact ⟨a * b, by simp⟩,
+        inv_mem' := by
+            rintro _ ⟨a, rfl⟩
+            exact ⟨a⁻¹, by simp⟩
+    }
+    -- We define a free K-module W with a basis indexed by
+    -- the elements of a double coset space which (in the totally
+    -- definite case) is finite)
+    let X := Doset.Quotient (Set.range (incl₁ F D)) U'
+    have h : Finite X :=
+        NumberField.FiniteAdeleRing.DivisionAlgebra.finiteDoubleCoset (K:=F) (D:=D) hU
+
+    letI := Fintype.ofFinite X
+    let m := Fintype.card X
+    let W := Fin m → K
+    let V := (WeightTwoAutomorphicFormOfLevel U K)
+
+    let index := (Fintype.equivFin X : X ≃ Fin m)
+    let rep : X → Dfx F D := Quot.out
+
+    -- We then define a linear map φ from V to the finite-dimensional space W.
+    -- V is a space of functions, and the map consists of evaluating
+    -- a function on representatives given by the rep function above.
+    let φ : V →ₗ[K] W :=
+        {
+        toFun := fun v i ↦ v.1.toFun (rep (index.symm i)),
+        map_add' := by
+            intros v₁ v₂
+            ext i
+            rfl
+        map_smul' := by
+            intros c v
+            ext i
+            rfl
+        }
+    -- Since we have a linear map φ from V to W and W is finite-dimensional,
+    -- it's enough to check that φ is injective.
```

Comment:
That is true. So instead of writing like a mathematician, why now now argue backwards like a computer scientist, and *reduce* the goal to checking that phi is injective?
```lean
  apply FiniteDimensional.of_injective φ
```
on line 78 just turns the goal immediately into the claim that phi is injective meaning that you don't need a `have` statement any more (just delete it) and you don't need to indent another two spaces (note that the code style for Lean code is indent 2 not indent 4).